src/scales/scale.time.js

```diff
@@ -358,31 +358,67 @@ function generate(min, max, capacity, options) {
 }
 
 /**
- * Returns the right and left offsets from edges in the form of {left, right}.
+ * Returns the right and left offsets from the edges of the chart in the form of {left, right}
+ * where each values is a relative width to the scale and ranges between 0 and 1.
+ * They add extra margins on the both sides by scaling down the original scale.
+ * Offsets are typically used for bar charts. They are calculated based on intervals
+ * between the data points to keep the leftmost and rightmost bars from being cut.
  * Offsets are added when the `offset` option is true.
  */
-function computeOffsets(table, ticks, min, max, options) {
+function computeOffsets(table, ticks, data, min, max, options) {
+	var pos = [];
+	var minInterval = 1;
+	var timeOpts = options.time;
+	var barThickness = options.barThickness;
 	var left = 0;
 	var right = 0;
-	var upper, lower;
-
-	if (options.offset && ticks.length) {
-		if (!options.time.min) {
-			upper = ticks.length > 1 ? ticks[1] : max;
-			lower = ticks[0];
-			left = (
-				interpolate(table, 'time', upper, 'pos') -
-				interpolate(table, 'time', lower, 'pos')
-			) / 2;
+	var i, ilen, curr, prev, length, width;
+
+	if (!options.offset) {
+		return {left: 0, right: 0};
+	}
+
+	data.forEach(function(timestamp) {
+		if (timestamp >= min && timestamp <= max) {
+			pos.push(interpolate(table, 'time', timestamp, 'pos'));
 		}
-		if (!options.time.max) {
-			upper = ticks[ticks.length - 1];
-			lower = ticks.length > 1 ? ticks[ticks.length - 2] : min;
-			right = (
-				interpolate(table, 'time', upper, 'pos') -
-				interpolate(table, 'time', lower, 'pos')
-			) / 2;
+	});
+
+	length = pos.length;
+	if (!length) {
+		return {left: 0, right: 0};
+	}
+
+	if (!barThickness) {
+		// Calculate minInterval
+		[data, ticks].forEach(function(timestamps) {
+			for (i = 0, ilen = timestamps.length; i < ilen; ++i) {
+				curr = interpolate(table, 'time', timestamps[i], 'pos');
+				minInterval = i > 0 ? Math.min(minInterval, curr - prev) : minInterval;
+				prev = curr;
+			}
+		});
+	}
+
+	if (!timeOpts.min) {
+		if (length === 1) {
```

Comment:
actually I just thought of one last thing. can we move setting `width` outside to occur just before the if statement? I think that would allow removing the duplicate code here and in the other if statement just below

Reply:
I'm not sure what you mean. `width` for the left and right offset needs to be calculated separately, and the calculation depends on the length of `pos` and `barThickness`.

Reply:
Yes, you're right. Nevermind